basis-library/integer/int-inf0.sml

```diff
@@ -1014,10 +1030,99 @@ structure IntInf =
+               else
+                  let
+                     val (_, ans) = word_tag_ans (I.quot (numi, deni))
+                  in 
+                     Prim.fromWord ans
+                  end
+            end
+
+         fun smallRem (num, den) =
+            let 
+               val (_, numi) = word_objptr_arg num
+               val (_, deni) = word_objptr_arg den
+               val (_, ans) = word_tag_ans (I.rem (numi, deni))
+            in 
+               Prim.fromWord ans
+            end
+
+         (*fun quotReserve (nlimbs, dlimbs) = reserve (S.- (nlimbs, dlimbs), 2)
+         fun remReserve (nlimbs, dlimbs) = reserve (dlimbs, 1)*)
       in
          val bigAdd = make (I.+!, Prim.+, S.max, 1)
          val bigSub = make (I.-!, Prim.-, S.max, 1)
          val bigMul = make (I.*!, Prim.*, S.+, 0)
+
+         fun bigQuot (num: bigInt, den: bigInt): bigInt =
+            if areSmall (num, den) then smallQuot (num, den)
+               else let
+                       val nlimbs = numLimbs num
+                       val dlimbs = numLimbs den
+                    in
+                       if S.< (nlimbs, dlimbs)
+                          then zero
+                          else if den = zero
+                                  then raise Div
+                                  else Prim.quot (num, den, reserve (S.- (nlimbs, dlimbs), 2))
+                    end
+
+         fun bigRem (num: bigInt, den: bigInt): bigInt =
+            if areSmall (num, den)
+               then smallRem (num, den)
+               else let 
+                       val nlimbs = numLimbs num
+                       val dlimbs = numLimbs den
+                    in 
+                       if S.< (nlimbs, dlimbs)
+                          then num
+                          else if den = zero
+                                  then raise Div
+                                  else Prim.rem (num, den, reserve (dlimbs, 1))
+                    end
+
+         fun bigQuotRem (num, den) =
+            if areSmall (num, den) then
+               (* the small versions are not optimized together,
+                * as this would not save a significant amount
+                * of time *)
+               (smallQuot (num, den), smallRem (num, den))
+            else  (* arguments are large *)
+               let
+                  val n_d_limbs as (nlimbs, dlimbs) =
+                     (numLimbs num, numLimbs den)
+               in
+                  (* try to avoid expensive operation in trivial cases *)
+                  if S.< n_d_limbs then
+                     (zero, num)
+                  else if den = zero then
+                     raise Div
+                  else  (* must perform operation *)
+                     let
+                        val qrs =
+                           Prim.quotRem (num, den,
+                                         reserve (S.- (nlimbs, dlimbs), 1),
+                                         reserve (dlimbs, 0))
```

Comment:
See discussion below, but you will need something more like:
```
Sz.+ (Sz.+ (Sz.* (Sz.zextdFromInt32 (Int32.quot (ObjptrWord.sizeInBits, 8)), 2),
      Sz.+ (bytesPerSequenceMetaData, (* Sequence MetaData *)
            case MLton.Align.align of (* alignment *)
               MLton.Align.Align4 => 0w3
              | MLton.Align.Align8 => 0w7)),
Sz.+ (reserve (S.- (nlimbs, dlimbs), 1),
      reserve (dlimbs, 0)))
```
because you need to compute the total bytes required to be available in the heap.